### DISCUSSION TITLE
Sharing tailwind between a design system and apps

### DISCUSSION THREAD

**Discussion by AndyOGo** [tailwindlabs/tailwindcss#8402] [Help] [answered]
We build a design system with [storybook](https://storybook.js.org/) to provide shared components.

Then we have several apps, which build on top of these components to implement user stories.

Now our problem is, how to share or reuse the CSS generated from tailwind for the design system across the apps?
And to use tailwind itself as part of the apps, in a manner that only classes which are not used in the design system get generated?

**Comment by nanorepublica** [comment, marked as answer]
Take a look at [Presets](https://tailwindcss.com/docs/presets) as a starting point, having a base config (or multiple configs) ought to let you scale across multiple apps along with the relevant component code packaged in to a package

**Reply by AndyOGo** [reply]
@nanorepublica Thank you very much! Presets work very well.

Now I have another problem.

Our component library creates a CSS file with a subset of tailwind classes and each app creates their own CSS file with a subset of tailwind classes.

Now if I load both, those subsets may overlap and I end up with duplicated class names, which may change the order of CSS classes and violate https://tailwindcss.com/blog/automatic-class-sorting-with-prettier#how-classes-are-sorted

I wonder how to handle this?

Shall I include the components from `node_modules` into my `content` config and skip the generated CSS from the component library?
I guess tailwind compiler isn't clever enough to follow `import` statements, right?

**Reply by nanorepublica** [reply]
I haven't done this myself (yet) so this will mostly be speculation. To answer your questions in reverse:

* Not sure about the tailwind compiler and import statements (I guess you mean JS import statements and not CSS import statements?). That might be worth a separate question/discussion
* Yeah, I would try to have each app compile the whole CSS required for that app to ensure the most optimum output which would avoid duplication and collisions. Including the components in `content` seems like a good starting point to me.

**Reply by AndyOGo** [reply]
@nanorepublica 
Thank you for your reply.

I re-read the docs again and found the related parts.

> If you’ve created your own reusable set of components that are styled with Tailwind and are importing them in multiple projects, make sure to configure Tailwind to scan those components for class names:
> `tailwind.config.js`
> ```js
> module.exports = {
>   content: [
>     './components/**/*.{html,js}',
>     './pages/**/*.{html,js}',
>     './node_modules/@my-company/tailwind-components/**/*.js',
>   ],
>   // ...
> }
> ```
https://tailwindcss.com/docs/content-configuration#working-with-third-party-libraries

> The way Tailwind scans your source code for classes is intentionally very simple — we don’t actually parse or execute any of your code in the language it’s written in, we just use regular expressions to extract every string that could possibly be a class name.
https://tailwindcss.com/docs/content-configuration#class-detection-in-depth
